background-processor/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+/// Rust-Lightning running properly, and (2) either can or should be run in the background. Its
+/// responsibilities are:
+/// * Monitoring whether the ChannelManager needs to be re-persisted to disk, and if so,
+///   writing it to disk/backups by invoking the callback given to it at startup.
+///   ChannelManager persistence should be done in the background.
+/// * Calling `ChannelManager::timer_chan_freshness_every_min()` every minute (can be done in the
+///   background).
+///
+/// Note that if ChannelManager persistence fails and the persisted manager becomes out-of-date,
+/// then there is a risk of channels force-closing on startup when the manager realizes it's
+/// outdated. However, as long as `ChannelMonitor` backups are sound, no funds besides those used
+/// for unilateral chain closure fees are at risk.
+pub struct BackgroundProcessor {
+	stop_thread: Arc<AtomicBool>,
+	/// May be used to retrieve and handle the error if `BackgroundProcessor`'s thread
+	/// panics.
+	pub thread_handle: JoinHandle<Result<(), std::io::Error>>
+}
+
+#[cfg(not(test))]
+const CHAN_FRESHNESS_TIMER: u64 = 60;
+#[cfg(test)]
+const CHAN_FRESHNESS_TIMER: u64 = 1;
+
+impl BackgroundProcessor {
+	/// Start a background thread that takes care of responsibilities enumerated in the top-level
+	/// documentation.
+	///
+	/// If `persist_manager` returns an error, then this thread will return said error (and `start()`
+	/// will need to be called again to restart the `BackgroundProcessor`). Users should wait on
+	/// [`thread_handle`]'s `join()` method to be able to tell if and when an error is returned, or
+	/// implement `persist_manager` such that an error is never returned to the `BackgroundProcessor`
+	///
+	/// `persist_manager` is responsible for writing out the `ChannelManager` to disk, and/or uploading
+	/// to one or more backup services. See [`ChannelManager::write`] for writing out a `ChannelManager`.
+	/// See [`FilesystemPersister::persist_manager`] for Rust-Lightning's provided implementation.
+	///
+	/// [`thread_handle`]: struct.BackgroundProcessor.html#structfield.thread_handle
+	/// [`ChannelManager::write`]: ../lightning/ln/channelmanager/struct.ChannelManager.html#method.write
+	/// [`FilesystemPersister::persist_manager`]: ../lightning_persister/struct.FilesystemPersister.html#impl
+	pub fn start<PM, ChanSigner, M, T, K, F, L>(persist_manager: PM, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>, logger: Arc<L>) -> Self
+	where ChanSigner: 'static + ChannelKeys + Writeable,
+	      M: 'static + chain::Watch<Keys=ChanSigner>,
+	      T: 'static + BroadcasterInterface,
+	      K: 'static + KeysInterface<ChanKeySigner=ChanSigner>,
+	      F: 'static + FeeEstimator,
+	      L: 'static + Logger,
+	      PM: 'static + Send + Fn(Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>,
```

Comment:
Is there any reason for this function to take an Arc by ownership instead of a `&ChannelManager`?
nit: ideally we wouldn't require a `ChannelManager` parameterized by Arcs for every type, though I'm not sure how you'd meed the required `'static` liftime without it so its not really a big deal.

Reply:
It's giving me this super helpful error atm: https://i.imgur.com/sD3T8A0.png

Reply:
Heh, yea, I'd seen that too. i believe its because the `callback` referenced is expecting the `Arc` still (cause what its calling is expecting the `Arc`).

Reply:
Hum even when I make `FilesystemPersister::persist_manager` expect a reference, still this error. I think what I need to do is specify the type of `node` in `callback`, so now trying to figure out how to specify type parameters in the `callback` closure...